File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int n, int k, vector<int> enemy) {
    int answer = enemy.size();
    priority_queue<int, vector<int>, less<int>> pq;
    
    for(int i = 0; i < enemy.size(); i++) {
        pq.push(enemy[i]);
        n -= enemy[i];
        
        if(n < 0) {
            if(k == 0) {
                answer = i;
                break;
            }
            
            // 무적권
            while(!pq.empty() && k > 0 && n < 0){
                k--;
                n += pq.top();
                pq.pop();
            }
        }
    }
    
    return answer;
}

File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.js
function solution(n, k, enemy) {
    let answer = enemy.length;
    
    let low = 0, high = enemy.length;
    while(low <= high) {
        let mid = Math.floor((low + high) / 2);
        let curEnemy = enemy.slice(0, mid).sort((a, b) => b - a);
        let curK = k;
        
        const curSum = curEnemy.reduce((acc, val) => {
            if (curK > 0) {
                curK--;
                return acc;
            }
            else
                return acc + val
        }, 0);
        
        if(n - curSum >= 0 && curK >= 0)
            low = mid + 1;
        else
            high = mid - 1;
    }
    
    answer = low - 1;
    
    return answer;
}